OSDOCS-19851: adds disconnected install docs RHCL (#112599)
openshift/openshift-docs: modules/proc-rhcl-disconnected-mirror-catalog.adoc

// Module included in the following assemblies:
//
// *install-guide/disconnected-install.adoc

:_mod-docs-content-type: PROCEDURE
[id="proc-rhcl-disconnected-mirror-catalog_{context}"]
= Mirroring the {prodname} Operator catalog

[role="_abstract"]
To begin your disconnected {prodname} installation, you must mirror the Red Hat Operator catalog to your disconnected environment. By mirroring, you are making a complete local copy of {prodname} on a private container registry that your isolated cluster can reach.

[NOTE]
====
Applying `ImageDigestMirrorSet` or `ImageTagMirrorSet` objects triggers `MachineConfig` object updates and restarts cluster nodes.
====

> .Prerequisites

Reviewer: I wonder if we should mention Istio-related images mirroring as a dependency anywhere? It is a hard RHCL dependency after all, right @mikenairn ?

Reviewer 2: It might be worth mention it, but i don't think we need to go into detail about it. This should be part of the mirroring of the cluster images themselves if we are using the default service mesh installed by the ClusterIngressOperator as i understand it.

Author: I think I understood this. Added a note. PTAL.

* You have a supported-version, disconnected {ocp} cluster.
* You configured a mirror registry.
* You have access to `registry.redhat.io` on an internet-connected workstation.
* You have credentials for your mirror registry configured in `${HOME}/.docker/config.json`.
* You have either an `ImageDigestMirrorSet` or `ImageTagMirrorSet` object configured for platform images.
* You installed the `oc-mirror` CLI tool for mirroring operator catalogs.

.Procedure

. On your internet-connected workstation, create an `ImageSetConfiguration` file named `rhcl-imageset-config.yaml` by using the following example:
+
.Example `ImageSetConfiguration` YAML
[source,yaml]
----
apiVersion: mirror.openshift.io/v2alpha1
kind: ImageSetConfiguration
mirror:
  operators:
  - catalog: registry.redhat.io/redhat/redhat-operator-index:v{product-version}
    packages:

>     - name: rhcl-operator <2>

Reviewer: `<2>` but there are only standard bullet points below, should it be a numbered list?

Author: whoops, not we don't support callouts anymore

  additionalImages:
  - name: registry.redhat.io/rhcl-1/coredns-rhel9:vX.Y.Z
----
+
* In the `mirror.operators.catalog` parameter, specify the Red Hat Operator catalog version matching your {ocp} cluster version.
* The value of the `mirror.operators.packages` parameter is the {prodname} Operator group. Authorino, Limitador, and DNS Operators are included automatically.
* The value of the `mirror.additionalImages.name` parameter is the CoreDNS image. Set this value if you want to use the Core DNS Operator. Replace `vX.Y.Z` with the version matching the {prodname} Operator release.

. Mirror the catalog to a local directory by running the following command:
+
[source,terminal]
----
$ oc mirror --v2 --config rhcl-imageset-config.yaml file://./rhcl-mirror
----
+
This command downloads the catalog and Operator images to the `rhcl-mirror` directory.

. Transfer the `rhcl-mirror` directory to your disconnected environment using removable media.

. On a host in the disconnected environment that has access to your mirror registry, push the images to the registry by running the following command:
+
[source,terminal,subs="+quotes"]
----

> $ $ oc mirror --v2 --config rhcl-imageset-config.yaml \

Reviewer: ```suggestion
$ oc mirror --v2 --config rhcl-imageset-config.yaml \
```

  --from file://./rhcl-mirror \

>   docker://_<mirror_registry>_:_<port>_

Reviewer: Not sure what's wrong here but `<port>` doesn't show up in cursive for me

Author: Huh, probably need to just enclose the whole value.

----
+
Replace `_<mirror_registry>_` and `_<port>_` with the path and port to your mirror registry.

> . Apply the `ImageDigestMirrorSet` or `ImageTagMirrorSet` to configure the cluster to use the mirrored images by running one of the following commands:

Reviewer: This should be _apply either or both_. `oc-mirror` generates IDMS and ITMS manifests based on how images are referenced in the mirrored operators, and I think we should generalize this here in case we have both occurrences in our images

.. Apply the `ImageDigestMirrorSet` by running the following command:
+
[source,terminal,subs="+quotes"]
----
$ oc apply -f results-*/cluster-resources/idms-oc-mirror.yaml
----

.. Apply the `ImageTagMirrorSet` by running the following command:
+
[source,terminal,subs="+quotes"]
----
$ oc apply -f results-*/cluster-resources/itms-oc-mirror.yaml
----

. Wait for the `MachineConfigPool` object to update by running the following command:
+
[source,terminal]
----
$ oc wait mcp --all --for=condition=Updated --timeout=30m
----
+
.Example output
[source,text]
----
machineconfigpool.machineconfiguration.openshift.io/master condition met
machineconfigpool.machineconfiguration.openshift.io/worker condition met
----

. Apply the generated `CatalogSource` manifest by running the following command:
+
[source,terminal,subs="+quotes"]
----
$ oc apply -f results-*/cluster-resources/catalogSource-*.yaml
----

> //Q: instead of the stars, can we give generic examples?

Reviewer: just the `oc apply -f cluster-resources/cs-redhat-operator-index-v4-21.yaml`

. Wait for the Operator catalog to be ready by running the following command:
+
[source,terminal]
----
$ oc wait catalogsource -n openshift-marketplace --all --for=condition=Ready --timeout=5m
----
+
.Example output
[source,text]
----

> catalogsource.operators.coreos.com/rhcl-operator-catalog-<timestamp> condition met

Reviewer: ```suggestion
catalogsource.operators.coreos.com/cs-redhat-operator-index-v4-21 condition met
```

* You configured a mirror registry.
* You have access to `registry.redhat.io` on an internet-connected workstation.
* You have credentials for your mirror registry configured in `${HOME}/.docker/config.json`.
* You have either an `ImageDigestMirrorSet` or `ImageTagMirrorSet` object configured for platform images.
* You installed the `oc-mirror` CLI tool for mirroring operator catalogs.

.Procedure

. On your internet-connected workstation, create an `ImageSetConfiguration` file named `rhcl-imageset-config.yaml` by using the following example:
+
.Example `ImageSetConfiguration` YAML
[source,yaml]
----
apiVersion: mirror.openshift.io/v2alpha1
kind: ImageSetConfiguration
mirror:
  operators:
  - catalog: registry.redhat.io/redhat/redhat-operator-index:v{product-version}
    packages:
    - name: rhcl-operator
  additionalImages:
  - name: registry.redhat.io/rhcl-1/coredns-rhel9:vX.Y.Z
----
+
* In the `mirror.operators.catalog` parameter, specify the Red Hat Operator catalog version matching your {ocp} cluster version.
* The value of the `mirror.operators.packages` parameter is the {prodname} Operator group. Authorino, Limitador, and DNS Operators are included automatically.

> * The value of the `mirror.additionalImages.name` parameter is the CoreDNS image. Set this value if you want to use the Core DNS Operator. Replace `vX.Y.Z` with the version matching the {prodname} Operator release.

Reviewer 2: This will be the dns-operator operator version. Not sure how easy it is to determine that, but we should make sure this obviously optional as most installs won't need it.

// Module included in the following assemblies:
//
// *install-guide/disconnected-install.adoc

:_mod-docs-content-type: PROCEDURE
[id="proc-rhcl-disconnected-mirror-catalog_{context}"]
= Mirroring the {prodname} Operator catalog

[role="_abstract"]
To begin your disconnected {prodname} installation, you must mirror the Red Hat Operator catalog to your disconnected environment. By mirroring, you are making a complete local copy of {prodname} on a private container registry that your isolated cluster can reach.

Applying `ImageDigestMirrorSet` or `ImageTagMirrorSet` objects triggers `MachineConfig` object updates and restarts cluster nodes.

[IMPORTANT]
====
{prodname} on {ocp} 4.19 and newer uses a dedicated Istio plugin to manage gateway traffic and does not require installation of {service-mesh}. However, the underlying Istio container images remain a hard dependency. In a disconnected environment, ensure that these plugin images are included in your image set configuration when mirroring the {prodname} Operator catalog.

> ====

Reviewer: I like the note 👍

Reviewer 2: This sounds a bit misleading to me. Assuming "{prodname}" here is RHCL, it sounds like you expect the user to add the images to the  RHCL image set configuration, but that wouldn't be the case since service mesh is part of the core OpenShift and should be mirrored as part of the process of creating a "disconnected" cluster.

I'm thinking more along the lines of:

> {prodname} on {ocp} 4.19 and newer automatically provisions Service Mesh via ClusterIngressOperator when you create the `openshift-default` GatewayClass. In disconnected environments, ensure the `redhat-operators` catalog is available (standard OpenShift requirement), no separate Service Mesh image mirroring needed.

@averevki Can you confirm this is the case?

Author: well you don't have to use service mesh with RHCL on OpenShift 4.19+ AFIAK, so if you don't use service mesh, do you have to mirror the Istio container images manually, or does the ingress operator still take care of that? maybe both cases here are true, depending on whether service mesh used?

Reviewer: > {prodname} on {ocp} 4.19 and newer automatically provisions Service Mesh via ClusterIngressOperator when you create the `openshift-default` GatewayClass. In disconnected environments, ensure the `redhat-operators` catalog is available (standard OpenShift requirement), no separate Service Mesh image mirroring needed.

I'm not sure if `redhat-operators` catalog would usually be available in the accurate disconnected installs, with only the specific operator and platform images from there being mirrored during the cluster installation. This is technically correct but I don't think we need to get into the specifics of Istio configurations here that much, just the generalized note for what user should pay attention to.

I think only the last sentence with `In a disconnected environment, ensure that these plugin images are included in your image set configuration when mirroring the {prodname} Operator catalog.` is indeed a bit misleading, but I think it is enough to rephrase it so Istio images should be a part of a _platform images mirroring_, and not the RHCL image set configuration, so user has the opportunity to investigate the Istio management method, and the related process to mirror the required images, himself. This is my take on the note:
```suggestion
[IMPORTANT]
====
{prodname} on {ocp} 4.19 and newer uses a dedicated Istio plugin to manage gateway traffic and does not require installation of {service-mesh}. However, the underlying Istio container images remain a hard dependency. In a disconnected environment, ensure that these images are mirrored alongside your cluster's image set configuration, as they are not included in the {prodname} Operator catalog.
====
```

Reviewer 2: This sounds fine to me. I don't think it need hold this up.
